### ISSUE TITLE
Root Cause of Missing Context

### ISSUE THREAD

**Issue by electrostat** [mapbox/mapbox-events-android#146]
Investigate missing `MapboxTelemetry.applicationContext` when generating a location event.

- possible causes could be a race condition or odd background state
- solution may be to just drop events when context is not set

Reference Ticket: https://github.com/mapbox/mapbox-events-android/pull/144

cc: @zugaldia

**Comment by electrostat** [comment]
After researching and investigating this situation, there is a big hole within our Telemetry sdk that could cause this crash. The hole results from event creation and the Telemetry Service from being decoupled from the `MapboxTelemetry` class. This means we can generate events (before pushing them) without `applicationContext` being set and `TelemetryService` could be generated and function without the context set. 

We already have a fix in that piggy-backs off of the ability for a service to get `applicationContext` to set `MapboxTelemetry.applicationContext` when it is null. So that is solved for us.

We need a solution that makes sure `applicationContext` is set when generating events. This is twofold here. This problem will take care of itself when we pull all of the event generation code out of our sdk and put it into the parent sdks for handling. There are some events (like AppUserTurnstile) and utilities (nav and telem) which will likely not be moving over. We will need a solution for those and a possible short-term solution for the other events. 

My recommendation is we make those utils and event creation dependent on `MapboxTelemetry`. A parent sdk/app could only receive the utils and event factories through a fully initialized MapboxTelemetry class.

**Comment by electrostat** [comment]
Researched #3 from above and all options require context to be used.

**Comment by electrostat** [comment]
I pulled in all the relevant tickets from previous instances of this context being null and causing issues. Putting them all here for easy access and review when solution has been made.

https://github.com/mapbox/mapbox-gl-native/issues/12748
https://github.com/mapbox/mapbox-events-android/pull/144
https://github.com/mapbox/mapbox-events-android/issues/201 -- will include in PR
https://github.com/mapbox/mapbox-events-android/issues/188
https://github.com/mapbox/mapbox-events-android/issues/181
https://github.com/mapbox/mapbox-events-android/pull/190
https://github.com/mapbox/mapbox-events-android/issues/170

**Comment by zugaldia** [comment]
Can we start with adding null check guards while we work on a longer term solution? I'd be interested in a fix able to make it to the upcoming Maps `f` release (due Sep 12).

**Comment by electrostat** [comment]
@zugaldia we already have extensive null checks to get to this point. I will investigate where else we can add them. Per https://github.com/mapbox/mapbox-gl-native/issues/12748 we already have a null check there when destroying the service.

**Comment by zugaldia** [comment]
I think you're referring to a different issue, https://github.com/mapbox/mapbox-gl-native/issues/12748 isn't a NPE on the service but on telemetry utils. There's no null check in the code I linked.

**Comment by andrlee** [comment]
@electrostat one of the options to consider is to replace all instances where `MapboxTelemetry.applicationContext` is called with `getApplicationContext()` calls in `TelemetryService` class. Please think about it and consider implementing a few integration tests that prove this theory.

**Comment by electrostat** [comment]
@andrlee 
1. Correct. Issue with service `Context` crashes is this exact issue.
2. This should work for everything within `TelemetryService`. We are only missing this in 1 or 2 places, since we migrated the majority with a check of `MapboxTelemetry.applicationContext`. Best to remove that check and move everything to `getApplicationContext()` exclusively.
3. I agree with this for a future refactor and changes to the current structure. If we want to work on implementing these changes across the sdk, I'm down with getting rid of the old way.

I will cut a ticket for this work and see what issues we run into. My only theory is there is some process we set off that uses a `TelemetryUtils` class that calls applicationContext. I don't see this ATM, but most likely failure point in this system.
